### PR TITLE
refactor(controller): do not short-circuit re-promotion to current freight

### LINKED ISSUE #1826: do not short-circuit re-promotion to stage's current freight
At present, if you promote a Stage to a piece of Freight that is _already_ its current piece of Freight, it will short-circuit and succeed immediately with no action taking place.

If the re-promotion is occurring because some sort of undesirable drift has occurred behind Kargo's back, this short-circuiting prevents Kargo from setting things right again. 😞 

We will stop this short-circuiting and permit the promotion mechanisms to fully do their job.

Some promotion mechanisms may, themselves, short-circuit. For instance, when writing to Git, if the repo/branch is found to _already_ be in the desired state, no new commit or PR is required. For a case like integration with an Argo CD App, there will be no short-circuiting and Kargo will cause Apps to refresh and sync.

cc @wanghong230

### PR DESCRIPTION
**Pull request by krancour**
Fixes #1826

@hiddeco I tapped into your re-verification process to trigger re-verification after a successful re-promotion.

### CODE REVIEW

File: internal/controller/promotions/promotions.go [hunk 1]
```diff
@@ -439,23 +433,34 @@ func (r *reconciler) promote(
 		// so we are safe from race conditions and can just update the status
 		// TODO: remove all patching of Stage status out of promo reconciler
 		if err = kubeclient.PatchStatus(ctx, r.kargoClient, stage, func(status *kargoapi.StageStatus) {
-			// control-flow Stage history is maintained in Stage controller.
-			// So we only modify history for normal Stages.
-			// (Technically, we should prevent creating promotion jobs on
-			// control-flow stages in the first place)
```
Comment (krancour) [inline internal/controller/promotions/promotions.go:445]: I'm removing this comment, which seems as if it was somewhat misplaced and just opening an issue for it instead.

Comment (krancour) [reply internal/controller/promotions/promotions.go:445]: Created #1830

File: internal/controller/promotions/promotions.go [hunk 2]
```diff
@@ -439,23 +433,34 @@ func (r *reconciler) promote(
...
-
 			if newStatus.Phase == kargoapi.PromotionPhaseSucceeded {
-				// Only push promotion to Stage status history if the promotion succeeded
-				status.Phase = kargoapi.StagePhaseVerifying
-				status.CurrentPromotion = nil
-				if stage.Spec.PromotionMechanisms != nil {
+				// Handle specific things that need to happen on success.
+				// 1. Trigger re-verification for re-promotions.
+				// 3. Otherwise, update the current freight and history.
+				// 2. Update the phase to Verifying and clear the current promotion.
```
Comment (hiddeco) [inline internal/controller/promotions/promotions.go:442]: Order seems off here.

File: internal/controller/promotions/promotions.go [hunk 3]
```diff
@@ -439,23 +433,34 @@ func (r *reconciler) promote(
...
+				if status.CurrentFreight != nil &&
+					status.CurrentFreight.Name == targetFreight.Name {
+					if err = kargoapi.ReverifyStageFreight(
```
Comment (hiddeco) [inline internal/controller/promotions/promotions.go:445]: While this absolutely does the job for now, I can't help but wonder if the Stage reconciler could become more self-determined.

File: internal/controller/promotions/promotions.go (hunk 2 above)
Comment (krancour) [reply internal/controller/promotions/promotions.go:442]: 🤦‍♂️ I swear I can count. Will fix.

File: internal/controller/promotions/promotions.go (hunk 3 above)
Comment (krancour) [reply internal/controller/promotions/promotions.go:445]: Agreed. @jessesuen and I have both wanted for a long time that each reconciler will only ever mutate the objects it reconciles and not touch others. That's a larger bit of refactoring that needs to happen at some point. Right now, my biggest concern is getting the functionality correct and then we can circle back to figure out how to clean things up a bit. I know you an I both have various improvements in the works that are chipping away at these things.

File: internal/controller/promotions/promotions.go (hunk 2 above)
Comment (krancour) [reply internal/controller/promotions/promotions.go]: Fixed in recent force push.

**Review by hiddeco — APPROVED** [review]

**Pull request merged**